src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -711,5 +711,212 @@ private static void ThrowMarkedCtorDoesNotTakeAllProvidedArguments()
+
+            object?[] constructorArguments = new object?[parameters.Length];
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                ref FactoryParameterContext parameter = ref parameters[i];
+                constructorArguments[i] = ((parameter.ArgumentIndex != -1)
+                    // Throws a NullReferenceException if arguments is null. Consistent with expression-based factory.
+                    ? arguments![parameter.ArgumentIndex]
+                    : GetService(
+                        serviceProvider,
+                        parameter.ParameterType,
+                        declaringType,
+                        parameter.HasDefaultValue)) ?? parameter.DefaultValue;
+            }
+
+            return invoker.Invoke(constructorArguments.AsSpan());
+        }
+
+        private static object ReflectionFactoryDirect(
+            ConstructorInvoker invoker,
+            IServiceProvider serviceProvider,
+            object?[]? arguments)
+        {
+            if (serviceProvider is null)
+                ThrowHelperArgumentNullExceptionServiceProvider();
+
+            if (arguments is null)
+                ThrowHelperNullReferenceException(); //AsSpan() will not throw NullReferenceException.
+
+            return invoker.Invoke(arguments.AsSpan());
+        }
+
+        /// <summary>
+        /// For consistency with the expression-based factory, throw NullReferenceException.
+        /// </summary>
+        [DoesNotReturn]
+        private static void ThrowHelperNullReferenceException()
+        {
+            throw new NullReferenceException();
+        }
+#elif NETSTANDARD2_1_OR_GREATER || NETCOREAPP
+        private static object ReflectionFactoryCanonical(
+            ConstructorInfo constructor,
+            FactoryParameterContext[] parameters,
+            Type declaringType,
+            IServiceProvider serviceProvider,
+            object?[]? arguments)
+        {
+            if (serviceProvider is null)
+                ThrowHelperArgumentNullExceptionServiceProvider();
+
+            object?[] constructorArguments = new object?[parameters.Length];
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                ref FactoryParameterContext parameter = ref parameters[i];
+                constructorArguments[i] = ((parameter.ArgumentIndex != -1)
+                    // Throws a NullReferenceException if arguments is null. Consistent with expression-based factory.
+                    ? arguments![parameter.ArgumentIndex]
+                    : GetService(
+                        serviceProvider,
+                        parameter.ParameterType,
+                        declaringType,
+                        parameter.HasDefaultValue)) ?? parameter.DefaultValue;
+            }
+
+            return constructor.Invoke(BindingFlags.DoNotWrapExceptions, binder: null, constructorArguments, culture: null);
+        }
+#endif
```

Comment:
Isn't this the same condition as the `elif` and the following? In other words, can't we just delete lines 912 and 914 now?

Reply:
It is the same as the same `elif` but also needs to compile for the `#if NET8_0_OR_GREATER`. It could be moved above in a compatible section.